felix/fv/pre_dnat_test.go

```diff
@@ -200,6 +206,87 @@ var _ = infrastructure.DatastoreDescribe("pre-dnat with initialized Felix, 2 wor
 				})
 
 				Context("with pre-DNAT policy to open pinhole to 8055", func() {
+					// Exec iptables-save, or iptables-save -t [table], if table is not "".
+					dumpIPTables := func(table string) []string {
+						execArgs := []string{"iptables-save"}
+						if table != "" {
+							execArgs = append(execArgs, "-t", table)
+						}
+						out, err := tc.Felixes[0].ExecOutput(execArgs...)
+						Expect(err).NotTo(HaveOccurred())
+						lines := strings.Split(out, "\n")
+						Expect(len(lines)).NotTo(Equal(1), "couldn't split on newline")
+						var grep []string
+						for _, line := range lines {
+							if strings.Contains(line, "cali-") && strings.Contains(line, "-A ") {
+								grep = append(grep, line)
+							}
+						}
+						return grep
+					}
+					dumpMangleTable := func() []string { return dumpIPTables("mangle") }
+					dumpFilterTable := func() []string { return dumpIPTables("filter") }
+
+					// Returns nil when the value of f() hasn't changed for the duration settledPeriod.
+					// Returns error if f() never settles up to the timeout.
+					// Immediately returns error if timeout is lower than settledPeriod.
+					waitToSettle := func(f func() int, settledPeriod, timeout time.Duration) error {
+						if timeout < settledPeriod {
+							return errors.New("Programmer error: timeout is less than settledPeriod")
+						}
+						var last int
+						var settledTimer, timeoutTimer *time.Timer
+						timeoutTimer = time.NewTimer(timeout)
+						for {
+							cur := f()
+
+							if settledTimer != nil {
+								if last != cur {
+									if !settledTimer.Stop() {
+										<-settledTimer.C
+									}
+									settledTimer = nil
+								} else {
+									select {
+									case <-settledTimer.C:
```

Comment:
With the way you're using the settledTimer, you're not actually using it as a timer at all!  You only select on it in a non-blocking way.  I think this code would be clearer if you just recorded the time since the last change and did 
```
last := f()
lastChange := time.Now()
for {
  cur := f()
  if cur != last {
    lastChange := time.Now()
  }
  last = cur
  if time.Since(lastChange) > settledPeriod {
    // we're done
  }
  ...
}
```